### PR TITLE
Improve TOPLEVEL_BINDING

### PR DESCRIPTION
**Pull request by chrisseaton**

### CODE REVIEW

File: spec/ruby/language/predefined/toplevel_binding_spec.rb [hunk 1]
```diff
@@ -0,0 +1,24 @@
+require File.expand_path('../../../spec_helper', __FILE__)
+
+describe "The TOPLEVEL_BINDING constant" do
+  before :each do
```
Comment (chrisseaton) [inline spec/ruby/language/predefined/toplevel_binding_spec.rb]: @eregon if you change this to `before :all`, I think you get errors in our logic for detecting slow specs

```
1)
An exception occurred during: before :all ERROR
NoMethodError: undefined method `mspec_old_ruby_exe' for #<MSpecEnv:0x121a>:MSpecEnv
/Users/chrisseaton/Documents/truffleruby/truffleruby/spec/truffle.mspec:216:in `method_missing'
/Users/chrisseaton/Documents/truffleruby/truffleruby/spec/truffle.mspec:216:in `send'
/Users/chrisseaton/Documents/truffleruby/truffleruby/spec/truffle.mspec:216:in `ruby_exe'
/Users/chrisseaton/Documents/truffleruby/truffleruby/spec/ruby/language/predefined/toplevel_binding_spec.rb:5:in `block (2 levels) in <top (required)>'
/Users/chrisseaton/Documents/truffleruby/truffleruby/src/main/ruby/core/enumerable.rb:478:in `block in all?'
/Users/chrisseaton/Documents/truffleruby/truffleruby/src/main/ruby/core/enumerable.rb:478:in `each'
/Users/chrisseaton/Documents/truffleruby/truffleruby/src/main/ruby/core/enumerable.rb:478:in `all?'
/Users/chrisseaton/Documents/truffleruby/truffleruby/spec/ruby/language/predefined/toplevel_binding_spec.rb:3:in `<top (required)>'
/Users/chrisseaton/Documents/truffleruby/truffleruby/src/main/ruby/core/kernel.rb:333:in `load'
/Users/chrisseaton/Documents/truffleruby/truffleruby/src/main/ruby/core/kernel.rb:333:in `load'
```

File: src/main/java/org/truffleruby/language/objects/shared/SharedObjects.java [hunk 2]
```diff
@@ -55,7 +55,7 @@ private static void shareContextRoots(RubyContext context) {
             stack.push(object);
         }
 
-        // Share all named modules and constants (including the shared TOPLEVEL_BINDING)
+        // Share all named modules and constants
```
Comment (chrisseaton) [inline src/main/java/org/truffleruby/language/objects/shared/SharedObjects.java]: What implications are there for sharing these new binding constants, @eregon? There's one created per top-level now.

File: spec/ruby/language/predefined/toplevel_binding_spec.rb [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+  it "does not include local variables defined in a required file" do
+    @binding_toplevel_variables.should_not include(:bar)
+  end
+
+  it "does not include local variables defined in eval" do
+    @binding_toplevel_variables.should_not include(:baz)
+  end
+
+  it "is always the same object for all top levels" do
+    @binding_toplevel_id.uniq.size.should == 1
```
Comment (eregon) [inline spec/ruby/language/predefined/toplevel_binding_spec.rb]: Any idea how to fix that one or what MRI/JRuby do?

File: spec/ruby/language/predefined/toplevel_binding_spec.rb (hunk 1 above)
Comment (eregon) [reply spec/ruby/language/predefined/toplevel_binding_spec.rb]: Is `ruby_exe` needed here or would `require` be enough?

File: src/main/java/org/truffleruby/language/SetTopLevelBindingNode.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+        try {
+            return child.execute(frame);
+        } finally {
+            if (previousValue == null) {
+                fields.removeConstant(context, this, "TOPLEVEL_BINDING");
+            } else {
+                fields.setConstant(context, this, "TOPLEVEL_BINDING", previousValue);
+            }
+        }
```
Comment (eregon) [inline src/main/java/org/truffleruby/language/SetTopLevelBindingNode.java]: So this re-set the TOPLEVEL_BINDING for every file loaded?
That's bad, as loading any file will deopt many constant lookups.
It also doesn't seem thread-safe.

File: spec/ruby/language/predefined/toplevel_binding_spec.rb [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+    @binding_toplevel_variables.should include(:foo)
+  end
+
+  it "does not include local variables defined in a required file" do
+    @binding_toplevel_variables.should_not include(:bar)
+  end
+
+  it "does not include local variables defined in eval" do
+    @binding_toplevel_variables.should_not include(:baz)
+  end
```
Comment (eregon) [inline spec/ruby/language/predefined/toplevel_binding_spec.rb]: Isn't simply that the TOPLEVEL_BINDING is built as if there was `TOPLEVEL_BINDING = binding` in the main script file (the filename passed to `ruby`)?
We had a node in the past setting up TOPLEVEL_BINDING once.

File: spec/ruby/language/predefined/toplevel_binding_spec.rb (hunk 1 above)
Comment (eregon) [reply spec/ruby/language/predefined/toplevel_binding_spec.rb]: It seems the main script file influences TOPLEVEL_BINDING so indeed we need `ruby_exe` then.

Comment (chrisseaton) [reply spec/ruby/language/predefined/toplevel_binding_spec.rb]: Yes these specs are all about top-level things, so I think it makes sense to isolate them.

File: spec/ruby/language/predefined/toplevel_binding_spec.rb (hunk 5 above)
Comment (chrisseaton) [reply spec/ruby/language/predefined/toplevel_binding_spec.rb]: If it's set in the main script then what is it set to for files loaded via `-rfoo`?

File: spec/ruby/language/predefined/toplevel_binding_spec.rb (hunk 3 above)
Comment (chrisseaton) [reply spec/ruby/language/predefined/toplevel_binding_spec.rb]: I think it's possibly a single frame for all files, but cleared of local variables for each new file. That's really tricky for us - we don't control creation of the frame!

File: src/main/java/org/truffleruby/language/SetTopLevelBindingNode.java (hunk 4 above)
Comment (chrisseaton) [reply src/main/java/org/truffleruby/language/SetTopLevelBindingNode.java]: Isn't there a global require lock?

Comment (chrisseaton) [reply src/main/java/org/truffleruby/language/SetTopLevelBindingNode.java]: And surely requiring any file is going to cause quite a lot of depots due to new classes and methods anyway?

File: spec/ruby/language/predefined/toplevel_binding_spec.rb (hunk 5 above)
Comment (eregon) [reply spec/ruby/language/predefined/toplevel_binding_spec.rb]: It seems the Binding is already created at that point but with no local variables:
```
$ ruby -r./sub main.rb
# in sub.rb
47001020681960
[]
# in main.rb
47001020681960
[:a, :b]
```

If we use `require_relative 'sub'` in main.rb, then locals of the main script (main.rb) are visible to both.
```
# in sub.rb
47090538696420
[:a, :b]
# in main.rb
47090538696420
[:a, :b]
```

File: spec/ruby/language/predefined/toplevel_binding_spec.rb (hunk 3 above)
Comment (chrisseaton) [reply spec/ruby/language/predefined/toplevel_binding_spec.rb]: Maybe we should create a single global `TOPLEVEL_BINDING` frame, and then in top level scopes we always assign into it as an outer scope?

File: src/main/java/org/truffleruby/language/SetTopLevelBindingNode.java (hunk 4 above)
Comment (eregon) [reply src/main/java/org/truffleruby/language/SetTopLevelBindingNode.java]: No, the require lock is per-file.
Yes, indeed, any file defining a new class/module or constant will invalidate some constant lookups (but could be under a namespace where the impact is much smaller than Object constants).

File: spec/ruby/language/predefined/toplevel_binding_spec.rb (hunk 3 above)
Comment (eregon) [reply spec/ruby/language/predefined/toplevel_binding_spec.rb]: That would not capture the locals of the main script though, isn't it?
I think we need to create the TOPLEVEL_BINDING early, and then patch its frame when just before executing the main script.

**Comment by chrisseaton** [comment]
I'll write some more specs first.

**Comment by eregon** [comment]
@chrisseaton Any progress on this?

File: src/main/java/org/truffleruby/language/SetTopLevelBindingNode.java [hunk 6]
```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright (c) 2018 Oracle and/or its affiliates. All rights reserved. This
```
Comment (wirthi) [inline src/main/java/org/truffleruby/language/SetTopLevelBindingNode.java]: In Graal.js there is a mandatory comma after the year. Graal seems to use the same. I don't think this has any consequence, just noting the difference.

**Comment by chrisseaton** [comment]
Closing to reduce the number of very stale PRs we have. We need to start this with a fresh set of good specs, anyhow.

**Pull request closed without merging**